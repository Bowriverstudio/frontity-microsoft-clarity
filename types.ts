import { Package } from "frontity/types";

/**
 * Analytics package to use Microsoft Clarity with Frontity.
 */
interface MicrosoftClarity extends Package {
  /**
   * Package name.
   */
  name: "microsoft-clarity";

  /**
   * The state exposed by this package.
   */
  state: {
    /**
     * State properties for the Microsoft Clarity  package.
     */
    microsoftClarity: {
      /**
       * Clarity tracking code.
       */
      trackingCode?: string;
    };
  };

}

export default MicrosoftClarity;
