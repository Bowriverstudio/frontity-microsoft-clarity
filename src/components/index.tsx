/* eslint-disable react/no-danger */
import { connect,Head } from "frontity";
import { Connect } from "frontity/types";
import React from "react";

import MicrosoftClarity from "../../types";

/**
 * Root component of the Microsoft Clarity package.
 *
 * It renders the Microsoft Clarity tag and set up each tracking ID defined in
 * the state.
 *
 * @remarks
 * This component is automatically rendered by Frontity and it's not meant to be
 * imported and used anywhere.
 *
 * @param props - Injected props by `connect`.
 *
 * @returns Root element.
 */
export const Root: React.FC<Connect<MicrosoftClarity>> = ({ state }) => {
  // Get Tracking ids from state.
  const { trackingCode } = state.microsoftClarity;

  // Render Tracker code.
  return (
    trackingCode.length > 0 && (
      <Head>
        <script>{`
(function(c,l,a,r,i,t,y){
  c[a]=c[a]||function(){(c[a].q=c[a].q||[]).push(arguments)};
  t=l.createElement(r);t.async=1;t.src="https://www.clarity.ms/tag/"+i;
  y=l.getElementsByTagName(r)[0];y.parentNode.insertBefore(t,y);
})(window, document, "clarity", "script", "${trackingCode}");
`}</script>
      </Head>
    )
  );
};

export default connect(Root);
