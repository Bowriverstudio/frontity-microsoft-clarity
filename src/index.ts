import MicrosoftClarity from "../types";
import Root from "./components";

const microsoftClarity: MicrosoftClarity = {
  name: "microsoft-clarity",
  roots: {
    microsoftClarity: Root,
  },
  state: {
    microsoftClarity: {},
  },
};

export default microsoftClarity;
