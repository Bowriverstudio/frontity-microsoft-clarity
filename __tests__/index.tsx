/**
 * @jest-environment node
 */

import { HelmetProvider } from "frontity";
import { State } from "frontity/types";
import React from "react";
import { FilledContext } from "react-helmet-async";
import TestRenderer from "react-test-renderer";

import microsoftClarity from "../src";
import { Root as MicrosoftClarity } from "../src/components";
import MicrosoftClarityPackage from "../types";

const getState = (): State<MicrosoftClarityPackage> => ({
  microsoftClarity: {
    trackingCode: "",
  },
});

describe("MicrosoftClarity", () => {
  test("works with a single tracking id", () => {
    const state = getState();

    state.microsoftClarity.trackingCode = "YOURCODE";

    const helmetContext = {};
    TestRenderer.create(
      <HelmetProvider context={helmetContext}>
        <MicrosoftClarity state={state} actions={null} />
      </HelmetProvider>
    ).toJSON();
    const head = (helmetContext as FilledContext).helmet;

    expect(head.script.toString()).toMatchSnapshot();
  });

  test("doesn't add anything if there's no tracking ids", () => {
    const state = getState();

    const helmetContext = {};
    TestRenderer.create(
      <HelmetProvider context={helmetContext}>
        <MicrosoftClarity state={state} actions={null} />
      </HelmetProvider>
    ).toJSON();
    const head = (helmetContext as FilledContext).helmet;

    expect(head.script.toString()).toMatchSnapshot();
  });

  // For Code coverage only.
  test("Empty definition", () => {
    expect(microsoftClarity).toEqual(microsoftClarity);
  });
});
