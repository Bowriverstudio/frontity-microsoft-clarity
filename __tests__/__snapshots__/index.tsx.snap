// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MicrosoftClarity doesn't add anything if there's no tracking ids 1`] = `""`;

exports[`MicrosoftClarity works with a single tracking id 1`] = `
"<script data-rh=\\"true\\" >
(function(c,l,a,r,i,t,y){
  c[a]=c[a]||function(){(c[a].q=c[a].q||[]).push(arguments)};
  t=l.createElement(r);t.async=1;t.src=\\"https://www.clarity.ms/tag/\\"+i;
  y=l.getElementsByTagName(r)[0];y.parentNode.insertBefore(t,y);
})(window, document, \\"clarity\\", \\"script\\", \\"YOURCODE\\");
</script>"
`;
